challenge_02/base_solution.py

```diff
@@ -7,17 +10,27 @@ class SelfAttention(nn.Module):
     One of the most popular part in NLP.
     If you understand self attention, you will understand the heard of transformers.
     """
-    def __init__(self):
+    def __init__(self, embedding_length: int) -> None:
         super().__init__()
-        self.W = None
+        self.e_length = embedding_length
+        self.TW = self._trainable_weights()
+
+    def _trainable_weights(self) -> Tuple[nn.Module, nn.Module, nn.Module]:
+        q = nn.Linear(self.e_length, self.e_length, bias=False)
+        k = nn.Linear(self.e_length, self.e_length, bias=False)
+        v = nn.Linear(self.e_length, self.e_length, bias=False)
+        return q, k, v
 
-    def forward(self, x):
-        q, k, v = self.W
-        # norm_attention = F.softmax(q*k)
-        # (norm_attention * v).sum(-1)
-        return
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        # Scaled Dot-Product Attention
+        q, k, v = self.TW
+        s = torch.matmul(q(x), torch.t(k(x)))
+        w = nn.functional.normalize(s)
```

Comment:
1. vyměň to za `F.normalize`
2. a jinak jsi tím nejsem jistý, v originálním článku se pro normalizaci používá `/sqrt(dim_k)`, z jakého zdroje jsi to bral

Reply:
ja som postupoval podla obrázku :D toho najslavnejšieho a tam bolo napísané normalize :D tak som našiel funkciu ktorá noramlizuje. Výhoda je že funkcia narozdiel od klasického podielu je blbomvzdorná. Ale máš pravdu podla paper tam `/sqrt(dim_k)` tak som to vymenil